src/orange/orangeinp/IntersectRegion.cc

```diff
@@ -1271,6 +1264,112 @@ bool Prism::encloses(Prism const& other) const
     return apothem_ >= other.apothem() && hh_ >= other.halfheight();
 }
 
+//---------------------------------------------------------------------------//
+// RevolvedSpecialTrapezoid
+//---------------------------------------------------------------------------//
+/*!
+ * Construct from a special trapezoid.
+ */
+RevolvedSpecialTrapezoid::RevolvedSpecialTrapezoid(SpecialTrapezoid&& trap)
+    : trap_(std::move(trap))
+{
+    CELER_VALIDATE(
+        trap_.bot().r[Bound::lo] >= 0 && trap_.top().r[Bound::lo] >= 0,
+        << "r values must be positive");
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Build surfaces.
+ *
+ * Surface are constructed by revolving each segment around the z axis. Thus:
+ * - segments parallel to z become z-aligned cylinders,
+ * - segments perpendicular to z become z-orthagonal planes,
+ * - other segments become z-aligned cones.
+ *
+ * If segment is coincident with the z axis, no surface is created, as it
+ * would enclose zero volume.
+ */
+void RevolvedSpecialTrapezoid::build(IntersectSurfaceBuilder& insert_surface) const
+{
+    auto const& bot = trap_.bot();
+    auto const& top = trap_.top();
+    constexpr auto left = Bound::lo;
+    constexpr auto right = Bound::hi;
+
+    // Create z planes first for faster short circuiting
+    if (trap_.variety() != SpecialTrapezoid::Variety::pointy_bot)
+    {
+        // Bottom z plane
+        insert_surface(Sense::outside, PlaneZ{bot.z});
+    }
+    if (trap_.variety() != SpecialTrapezoid::Variety::pointy_top)
+    {
+        // Top z plane
+        insert_surface(Sense::inside, PlaneZ{top.z});
+    }
+
+    SoftClose soft_close(trap_.abs_tol());
+
+    // Lambda for creating a cylindrical/conical surface, or no surface
+    auto make_vertical_surface
+        = [&](real_type r_bot, real_type r_top, Sense sense) {
+              if (!soft_close(r_bot, r_top))
+              {
+                  // Conical surface
+                  insert_surface(
+                      sense, this->make_cone({r_bot, bot.z}, {r_top, top.z}));
+              }
+              else if (!soft_close(real_type{0}, r_bot))
+              {
+                  // Cylindrical surface
+                  insert_surface(sense, CCylZ(r_bot));
+              }
```

Comment:
For clarity I would also write
```suggestion
              }
              else
              {
                  // r_top = r_bottom = 0: do not create a surface
              }
```